src/Domain/DonationRepository.php

```diff
@@ -667,24 +649,69 @@ public function removeAllFundingWithdrawalsForDonation(Donation $donation): void
         });
     }
 
-    private function setSalesforceIdIfNeeded(Donation $donation, Salesforce18Id $salesforceId): void
+    /**
+     * Re-queues proxy objects to Salesforce en masse.
+     *
+     * By using FIFO queues and deduplicating on UUID if there are multiple consumers, we should make it unlikely
+     * that Salesforce hits Donation record lock contention issues.
+     *
+     * @param RoutableMessageBus $bus   We had to not DI this to avoid a circular dependency.
+     * @return int  Number of objects pushed
+     */
+    public function pushSalesforcePending(\DateTimeImmutable $now, RoutableMessageBus $bus): int
```

Comment:
I think I remember couldn't inject the bus via the constructor because that would make a circular dependency, but I think we could probably inject it via a setter as we do for some other dependencies. Maybe simpler than injecting just to this method? 

Also we could widen the type to `MessageBusInterface`.

Reply:
I think because it's a Doctrine repo, DI in the constructor which has to follow a very specific signature isn't really possible? I'm not really sure about the accuracy of my circular deps comment, I think possibly that was incorrectly migrated from when the logic lived somewhere else.

Agree about widening the type, done in c0d6d320d245fad4e77560e4cc4ff68e6d4fc19a. I'm not sure about adding a method to set the service if exactly one thing uses it, seems a bit like adding complexity earlier than it's helpful to me.